experiments: Add taint labels (#698)
returntocorp/semgrep-docs: docs/experiments/taint-labels.md

---
slug: taint-labels
append_help_link: true
description: "Taint labels increase the expressiveness of taint-mode by allowing to specify and track different kinds of taint through labels."
---

# Taint labels

> Previously taint mode could only track one kind of taint, data was either tainted or clean and that had to be specified using no more than a single pattern operator. Sometimes this is not enough. For example, data may become _dangerous_ through several steps that are not easily specified with a single pair source-sink. Taint labels increase the expressiveness of taint-mode by allowing to specify and track different kinds of taint through labels. 

Author: > Can we say that: `Taint labels allow you to match more tainted sources`.?

I find it a bit confusing to use "match" here. The matching expressiveness didn't change because for that we use the matching engine. We could perhaps say `Taint labels allow you to specify more complex taint sources`.

> We can attach labels to taint sources with the `label` key, e.g. `label: TAINTED`. Any valid Python identifier is accepted as a label. With that, the taint produced by a source will be labeled accordingly and can be individually tracked. A sink can be restricted to a subset of labels, using the `requires` key and a Python Boolean expression over labels, e.g. `requires: LABEL1 and not LABEL2`. A source can also be restricted via the `requires` key in the same way as a sink, in which case extra taint will only be produced if the source itself is tainted an satisfies the `requires` formula.

Author: > EVIL and INPUT are not valid Python identifiers, right?

I would say it's not right, have you tried running a Python REPL and creating a variable named `INPUT`? Works for me.

> We can attach labels to taint sources with the `label` key, e.g. `label: TAINTED`. Any valid Python identifier is accepted as a label. With that, the taint produced by a source will be labeled accordingly and can be individually tracked. A sink can be restricted to a subset of labels, using the `requires` key and a Python Boolean expression over labels, e.g. `requires: LABEL1 and not LABEL2`. A source can also be restricted via the `requires` key in the same way as a sink, in which case extra taint will only be produced if the source itself is tainted an satisfies the `requires` formula.

Author: > do you really mean user-defined identifiers or Python reserved words?

I mean user-defined identifiers. I would need to understand why you are saying that `INPUT` isn't a valid one. It is as far as I can tell. https://docs.python.org/3/reference/lexical_analysis.html#identifiers

> "A sink can be restricted to a subset of labels" - Does it mean something as following: Limit combinations of labels that need to be included or excluded using required key?

I think yes. Typically taint reaching a sink triggers a finding. Now you can restrict _what_ taint (what labels) cannot go into that sink and therefore will trigger a finding. Perhaps taint A is safe wrt this sink, but taint B is dangerous, then you use `requires: B`. Maybe B is dangerous but only if it's not also A, because A is some sort of sanitizer, then you use `requires: B and not A`, etc.

> Python Boolean expressions (in your text) and Boolean operators are the same things? (Just asking to learn more myself.) https://note.nkmk.me/en/python-boolean-operation/

No, `A and not B` is an expression, whereas the Boolean operators used in that expression are `and` and `not`. You could however say that you can combine labels using Python Boolean operators.

> "A source can also be restricted via the requires key in the same way as a sink, in which case extra taint will only be produced if the source itself is tainted an satisfies the requires formula." - I am not sure I understand this whole sentence... what is the "extra" taint?

Maybe "extra taint" isn't the best term here. Given `foo(data)`, typically you could say that `foo(...)` is a taint source, that is, the output of `foo` is tainted. Now you can say:

```
- pattern: foo(...)
  requires: A
  label: B
```

This means that `foo` is a taint source, yes, but *only* when it receives data that is already tainted with label `A`, in which case `foo` will add the label `B` to that data. So the data is already tainted, and `foo` adds an extra label to it.

> For example, let's say that `user_input` is dangerous but only when it passes through the `evil` function. We could specify this with taint labels as follows:

Reviewer: ```suggestion
In the example below, let's say that `user_input` is dangerous but only when it passes through the `evil` function. This can be specified with taint labels as follows:
```
